libtextsecure/message_receiver.js

```diff
@@ -717,11 +717,70 @@ MessageReceiver.prototype.extend({
       deviceId: parseInt(textsecure.storage.user.getDeviceId(), 10),
     };
 
+    let conversation;
+    try {
+      conversation = await window.ConversationController.getOrCreateAndWait(envelope.source, 'private');
+    } catch (e) {
+      window.log.info('Error getting conversation: ', envelope.source);
+    }
+    const getCurrentSessionBaseKey = async () => {
+      const record = await sessionCipher.getRecord(address.toString());
+      if (!record)
+        return null;
+      const openSession = record.getOpenSession();
+      if (!openSession)
+        return null;
+      const { baseKey } = openSession.indexInfo;
+      return baseKey
+    };
+    const captureActiveSession = async () => {
+      this.activeSessionBaseKey = await getCurrentSessionBaseKey(sessionCipher);
+    };
+    const restoreActiveSession = async () => {
+      const record = await sessionCipher.getRecord(address.toString());
+      if (!record)
+        return
+      record.archiveCurrentState();
+      const sessionToRestore = record.sessions[this.activeSessionBaseKey];
+      record.promoteState(sessionToRestore);
+      record.updateSessionState(sessionToRestore);
+      await textsecure.storage.protocol.storeSession(address.toString(), record.serialize());
+    };
+    const deleteAllSessionExcept = async (sessionBaseKey) => {
+      const record = await sessionCipher.getRecord(address.toString());
+      if (!record)
+        return
+      const sessionToKeep = record.sessions[sessionBaseKey];
+      record.sessions = {}
+      record.updateSessionState(sessionToKeep);
+      await textsecure.storage.protocol.storeSession(address.toString(), record.serialize());
+    };
+    const handleSessionReset = async () => {
+      const currentSessionBaseKey = await getCurrentSessionBaseKey(sessionCipher);
+      // console.warn('%cdecipher session %s', 'color:red;', currentSessionBaseKey);
```

Comment:
Intentional?

Reply:
Half-intentional 🤔 
Wanted to keep a record of where to print the session, in case someone else wanted to start printing it. But I should just keep it locally as a stash.